openstack_compute_instance_v2.master: diffs didn't match during apply. This is a bug with Terraform and should be reported.

Reporting as indicated by the error.

I am trying to provision a new instance from a volume instantiated from a source volume id.

```
resource "openstack_blockstorage_volume_v1" "myvol" {
  name = "myvol"
  size = "50"
  source_vol_id = "d35aad44-9af9-4047-9766-049a17d66623"
}

resource "openstack_compute_instance_v2" "myinstance" {
  name = "myinstance"
  flavor_id = "1601"

  block_device = {
    uuid = "${openstack_blockstorage_volume_v1.myvol.0.id}"
    source_type = "volume"
    destination_type = "volume"
    boot_index = "0"
    volume_size = "50"
  }

  network = { uuid = "55ed6f55-a305-4073-b5a3-99bd453e866c" }
}
```

First run of ```terraform apply``` results in the error mentioned in the issue title. If I run it a second time, the instance is created successfully.

```terraform apply``` output: https://gist.github.com/jfchevrette/1e2ee15762efbd4df1a7
diff reason from TF_LOG=debug: https://gist.github.com/jfchevrette/f81d39464cc43c98f87e

@phinze I've figure out the fix for this, but I don't understand why  :smile: 

This issue only happens when the `block_device.uuid` attribute is set to a variable / exported value. In that situation, the following error is seen:

```
2015/12/03 03:58:22 [ERROR] openstack_compute_instance_v2.foo: diffs didn't match
2015/12/03 03:58:22 [ERROR] openstack_compute_instance_v2.foo: reason: extra attributes: block_device.0.delete_on_termination, block_device.0.boot_index, block_device.#, block_device.0.uuid, block_device.0.source_type, block_device.0.destination_type, block_device.0.volume_size
```

I'm able to fix this by creating a "proper" set function like this:

```golang
func resourceComputeBlockDeviceHash(v interface{}) int {
  var buf bytes.Buffer                                  
  m := v.(map[string]interface{})                       
  buf.WriteString(fmt.Sprintf("%s", m["uuid"].(string)))
                                                        
  return hashcode.String(buf.String())                  
}                                                       
```

Rather than the one I stole from `resource_aws_instance.go`:

```golang
Set: func(v interface{}) int {                                            
  // there can only be one bootable block device; no need to hash anything
  return 0                                                                
},                                                                        
```

Why doesn't the above work in this situation? I read it as saying "no matter what values, just return 0 as a hashcode", so regardless of any changing values, it'd all be grouped under the same hashed set.

I need to make this change to `block_device` to support the request in #4042 anyway (funny how these things work out), but I'd like to know why the current code causes this particular issue.

--- a/builtin/providers/openstack/resource_openstack_compute_instance_v2.go
+++ b/builtin/providers/openstack/resource_openstack_compute_instance_v2.go
@@ -176,12 +176,7 @@ func resourceComputeInstanceV2() *schema.Resource {
 				ForceNew: true,
 			},
 			"block_device": &schema.Schema{
-				// TODO: This is a set because we don't support singleton
-				//       sub-resources today. We'll enforce that the set only ever has
-				//       length zero or one below. When TF gains support for
-				//       sub-resources this can be converted.
-				//       As referenced in resource_aws_instance.go
-				Type:     schema.TypeSet,
+				Type:     schema.TypeList,
 				Optional: true,
 				ForceNew: true,
 				Elem: &schema.Resource{
@@ -212,10 +207,6 @@ func resourceComputeInstanceV2() *schema.Resource {
 							Default:  false,
 						},
 					},
-				},
-				Set: func(v interface{}) int {
-					// there can only be one bootable block device; no need to hash anything
-					return 0
 				},
 			},
 			"volume": &schema.Schema{
@@ -352,9 +343,8 @@ func resourceComputeInstanceV2Create(d *schema.ResourceData, meta interface{}) e
 		}
 	}
 
-	if v, ok := d.GetOk("block_device"); ok {
-		vL := v.(*schema.Set).List()
-		for _, v := range vL {
+	if vL, ok := d.GetOk("block_device"); ok {
+		for _, v := range vL.([]interface{}) {
 			blockDeviceRaw := v.(map[string]interface{})
 			blockDevice := resourceInstanceBlockDeviceV2(d, blockDeviceRaw)
 			createOpts = &bootfromvolume.CreateOptsExt{
@@ -1239,9 +1229,8 @@ func checkVolumeConfig(d *schema.ResourceData) error {
 		}
 	}
 
-	if v, ok := d.GetOk("block_device"); ok {
-		vL := v.(*schema.Set).List()
-		if len(vL) > 1 {
+	if vL, ok := d.GetOk("block_device"); ok {
+		if len(vL.([]interface{})) > 1 {
 			return fmt.Errorf("Can only specify one block device to boot from.")
 		}
 	}

--- a/builtin/providers/openstack/resource_openstack_compute_instance_v2_test.go
+++ b/builtin/providers/openstack/resource_openstack_compute_instance_v2_test.go
@@ -253,9 +253,9 @@ func TestAccComputeV2Instance_multi_secgroups(t *testing.T) {
 	})
 }
 
-func TestAccComputeV2Instance_bootFromVolume(t *testing.T) {
+func TestAccComputeV2Instance_bootFromVolumeImage(t *testing.T) {
 	var instance servers.Server
-	var testAccComputeV2Instance_bootFromVolume = fmt.Sprintf(`
+	var testAccComputeV2Instance_bootFromVolumeImage = fmt.Sprintf(`
 		resource "openstack_compute_instance_v2" "foo" {
 			name = "terraform-test"
 			security_groups = ["default"]
@@ -276,7 +276,46 @@ func TestAccComputeV2Instance_bootFromVolume(t *testing.T) {
 		CheckDestroy: testAccCheckComputeV2InstanceDestroy,
 		Steps: []resource.TestStep{
 			resource.TestStep{
-				Config: testAccComputeV2Instance_bootFromVolume,
+				Config: testAccComputeV2Instance_bootFromVolumeImage,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckComputeV2InstanceExists(t, "openstack_compute_instance_v2.foo", &instance),
+					testAccCheckComputeV2InstanceBootVolumeAttachment(&instance),
+				),
+			},
+		},
+	})
+}
+
+func TestAccComputeV2Instance_bootFromVolumeVolume(t *testing.T) {
+	var instance servers.Server
+	var testAccComputeV2Instance_bootFromVolumeVolume = fmt.Sprintf(`
+	  resource "openstack_blockstorage_volume_v1" "foo" {
+			name = "terraform-test"
+			size = 5
+			image_id = "%s"
+		}
+
+		resource "openstack_compute_instance_v2" "foo" {
+			name = "terraform-test"
+			security_groups = ["default"]
+			block_device {
+				uuid = "${openstack_blockstorage_volume_v1.foo.id}"
+				source_type = "volume"
+				volume_size = 5
+				boot_index = 0
+				destination_type = "volume"
+				delete_on_termination = true
+			}
+		}`,
+		os.Getenv("OS_IMAGE_ID"))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckComputeV2InstanceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccComputeV2Instance_bootFromVolumeVolume,
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckComputeV2InstanceExists(t, "openstack_compute_instance_v2.foo", &instance),
 					testAccCheckComputeV2InstanceBootVolumeAttachment(&instance),

